requests-for-comments/2024-07-01-alerting.md

```diff
@@ -0,0 +1,160 @@
+
+**Chunks**: Divide large lists of tasks into smaller, more manageable chunks. This method is particularly useful for 
+handling bulk operations without overwhelming the system. Chunks ensure that large sets of alert checks are processed 
+in smaller batches, which are then sequentially handled by chains within groups. Celery will make sure to not launch a 
+full task for each item in the list, but rather a task for each chunk.
+
+Combining these three constructs (groups, chains, and chunks) provides a robust framework for managing the complexity 
+of alert checks and notifications. This approach ensures scalability, efficiency, and maintainability, making it 
+suitable for handling potentially large volumes of alerts and checks in PostHog.
+
+##### Check Alert Task
+
+The `check_alert` task evaluates an alert condition and logs the result. If the condition is met, it triggers a 
+notification task.
+
+Rough code outline:
+
+```python
+from celery import shared_task, chain
+from .models import Alert, AlertCheck
+from django.utils import timezone
+
+@shared_task # Think about expiration time and timeout
+def check_alert(alert_id):
+    alert = Alert.objects.get(id=alert_id)
+    
+    # Keep in mind idempotency in case of retries - e.g., check if check is already in the database for this interval
+
+    calculated_value = 42  # Example calculated value - hand off calculation to existing insight code
+    threshold_met = True  # Example threshold check
+
+    alert_check = AlertCheck.objects.create(
+        alert=alert,
+        calculated_value=calculated_value,
+        threshold_met=threshold_met,
+        ...
+    )
+
+    if threshold_met:
+        send_notification.s(alert_check.id).delay() #  Launch notification task based on check object
+
+    return alert.id
+```
+
+##### Send Notification Task
+
+The `send_notification` task handles sending notifications and updates the `AlertCheck` with the notification status.
+
+```python
+@shared_task # Think about expiration time and timeout
+def send_notification(alert_check_id):
+    alert_check = AlertCheck.objects.get(id=alert_check_id)
+
+    # Keep in mind idempotency in case of retries - e.g., check if notification was already sent
+
+    success = True  # Assume success
+    notification_status = {"status": "success"}
+
+    ...  # Send notification logic
+
+    alert_check.notification_sent = True
+    alert_check.save()
+
+    return success
+```
+
+##### Scheduling Alert Checks
+
+The `schedule_alert_checks` task runs at regular intervals, creating groups and chains of alert check tasks to manage 
+concurrency and sequence.
+
+```python
+from celery import group, chain
+
+@shared_task
+def schedule_alert_checks():
+    alerts = Alert.objects.all()
+    alert_ids = [alert.id for alert in alerts]
+
+    alert_id_groups = [alert_ids[i:i+10] for i in range(0, len(alert_ids), 10)]
+
+    group_of_chains = group(
+        chain(check_alert.chunks(group, 10)) for group in alert_id_groups
```

Comment:
when we do implement this it definitely needs a comment - i think i understand what it's doing but i bet i'm wrong 🙈

Reply:
That's my concern as well, it looks a bit difficult, and that's why I suggest for + limit_concurrency